collect/central_collector.go

```diff
@@ -119,18 +153,235 @@ func (c *CentralCollector) collect() {
 
 }
 
-func (c *CentralCollector) processSpan(sp *types.Span) {
-	// add the span to the cache
-	trace := c.SpanCache.Get(sp.TraceID)
-	if trace == nil {
-		err := c.SpanCache.Set(sp)
-		if err != nil {
-			c.Logger.Error().Logf("error adding span with trace ID %s to cache: %s", sp.TraceID, err)
-			return
+func (c *CentralCollector) processTraces() error {
+	for {
+		select {
+		case <-c.done:
+			return nil
+		default:
+			ids := c.SpanCache.GetTraceIDs(processTracesBatchSize)
+			if len(ids) == 0 {
+				continue
+			}
+
+			statuses, err := c.Store.GetStatusForTraces(context.Background(), ids)
+			if err != nil {
+				c.Logger.Error().Logf("error getting statuses for traces: %s", err)
+				continue
+			}
+			for _, status := range statuses {
+				switch status.State {
+				case centralstore.DecisionKeep:
+					c.send(status)
+					c.SpanCache.Remove(status.TraceID)
+
+				case centralstore.DecisionDrop:
+					c.SpanCache.Remove(status.TraceID)
+				default:
+					c.Logger.Debug().Logf("trace %s is still pending", status.TraceID)
+				}
+			}
+		}
+
+		select {
+		case <-c.done:
+			return nil
+		default:
+		}
+
+		ticker := c.Clock.NewTicker(processTracesBackoffInterval)
+		select {
+		case <-c.done:
+			return nil
```

Comment:
If you do need a ticker for reasons I don't understand yet, you need to Stop it here too.